neqo-common/src/udp.rs

```diff
@@ -0,0 +1,155 @@
+        let socket = std::net::UdpSocket::bind(addr)?;
+
+        Ok(Self {
+            state: quinn_udp::UdpSocketState::new((&socket).into())?,
+            socket: tokio::net::UdpSocket::from_std(socket)?,
+        })
+    }
+
+    /// See [`tokio::net::UdpSocket::local_addr`].
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.socket.local_addr()
+    }
+
+    /// See [`tokio::net::UdpSocket::writable`].
+    pub async fn writable(&self) -> Result<(), io::Error> {
+        self.socket.writable().await
+    }
+
+    /// See [`tokio::net::UdpSocket::readable`].
+    pub async fn readable(&self) -> Result<(), io::Error> {
+        self.socket.readable().await
+    }
+
+    /// Send the UDP datagram on the specified socket.
+    pub fn send(&self, d: Datagram) -> io::Result<usize> {
+        let transmit = Transmit {
+            destination: d.destination(),
+            ecn: EcnCodepoint::from_bits(Into::<u8>::into(d.tos())),
+            contents: d.into_data().into(),
+            segment_size: None,
+            src_ip: None,
+        };
+
+        let n = self.socket.try_io(Interest::WRITABLE, || {
+            self.state
+                .send((&self.socket).into(), slice::from_ref(&transmit))
+        })?;
+
+        assert_eq!(n, 1, "only passed one slice");
+
+        Ok(n)
+    }
+
+    /// Receive a UDP datagram on the specified socket.
+    pub fn recv(&self, local_address: &SocketAddr) -> Result<Option<Datagram>, io::Error> {
+        let mut buf = [0; u16::MAX as usize];
+
+        let mut meta = RecvMeta::default();
+
+        match self.socket.try_io(Interest::READABLE, || {
+            self.state.recv(
+                (&self.socket).into(),
+                &mut [IoSliceMut::new(&mut buf)],
+                slice::from_mut(&mut meta),
+            )
+        }) {
+            Ok(n) => {
+                assert_eq!(n, 1, "only passed one slice");
+            }
+            Err(ref err)
+                if err.kind() == io::ErrorKind::WouldBlock
+                    || err.kind() == io::ErrorKind::Interrupted =>
+            {
+                return Ok(None)
+            }
+            Err(err) => {
+                return Err(err);
+            }
+        };
+
+        if meta.len == 0 {
+            eprintln!("zero length datagram received?");
+            return Ok(None);
+        }
+
+        if meta.len == buf.len() {
+            eprintln!("Might have received more than {} bytes", buf.len());
+        }
+
+        Ok(Some(Datagram::new(
+            meta.addr,
+            *local_address,
+            meta.ecn.map(|n| IpTos::from(n as u8)).unwrap_or_default(),
+            None, // TODO: get the real TTL https://github.com/quinn-rs/quinn/issues/1749
+            &buf[..meta.len],
+        )))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{IpTos, IpTosDscp, IpTosEcn};
+
+    #[tokio::test]
+    async fn datagram_tos() -> Result<(), io::Error> {
```

Comment:
This test doesn't seem to terminate on Windows.

Reply:
You can see on lines 138 and 141 that there is a `.await`.  That means that this needs to be async.  Or rather, it is easier to use `#[tokio::test]` rather than `#[test]` so that this runs faster.

As for the termination, it's possible that port 8080 is taken.  I believe that you can bind the receiver to a random/ephemeral port instead, which might help this complete.  No idea whether that is the problem though.

Reply:
Found the problem. Setting ECN CE is not allowed on Windows.

With https://github.com/mozilla/neqo/pull/1604/commits/e9ba397891451563e4c5674a0949ee060e8a83d7 we are now using ECN ECT1 for testing.

> PASS [   0.236s] neqo-common udp::tests::datagram_tos

---



> As for the termination, it's possible that port 8080 is taken. I believe that you can bind the receiver to a random/ephemeral port instead, which might help this complete. No idea whether that is the problem though.

Thanks Martin. Wasn't the issue here. Still, good idea. Changed with https://github.com/mozilla/neqo/pull/1604/commits/8ced9d9b97e39fb7e8d5e8d01c251f54384c7eda.